packages/op-datasets/src/op_datasets/etl/intermediate/testutils.py

```diff
@@ -0,0 +1,170 @@
+from .udfs import create_duckdb_macros
+
+log = structlog.get_logger()
+
+
+class IntermediateModelTestBase(unittest.TestCase):
+    """Base Class for Intermediate Model Unit Tests.
+
+    This class helps with fetching and locally storing sample data for use in intermediate
+    model unit tests.
+
+    The test data is stored in a local duckdb file. If the file does not exist the data is
+    fetched from GCS and stored so it can be used by subsequent runs.
+
+    Fetching is disabled by default. To enable fetching set `_enable_fetching = True` on
+    the child class. This should be done explicitly when you wish to fetch data for the
+    first time or when you want to update the existing test data.
+
+    Users of the class must supply a "block_filters" to narrow down the input data to
+    a set of blocks. This helps control the test data file size.
+    """
+
+    # Input parameters (must be set by child class)
+
+    # The path where input data will be stored.
+    inputdata: InputTestData
+
+    # Chains that should be included in the test data.
+    chains: list[str]
+
+    # Date that should be included in the test data.
+    dateval: date
+
+    # Datasets that should be included in the test data.
+    datasets: list[str]
+
+    # Specify blocks that should be included in the test data.
+    block_filters: list[str]
+
+    # Internal duckdb_client
+    _enable_fetching = False
+    _duckdb_client = None
+
+    @classmethod
+    def setUpClass(cls):
+        """Set up the test case.
+
+        If the duckdb file does not exist yet this method fetches the data from GCS and
+        stores it locally.
+
+        This method executes the model under test and creates temporary tables in DuckDB
+        with the model results.
+        """
+        load_model_definitions()
+
+        db_path = cls.inputdata.path(f"testdata/{cls.__name__}.duck.db")
+        os.makedirs(os.path.dirname(db_path), exist_ok=True)
+
+        cls._duckdb_client = duckdb.connect(db_path)
+        create_duckdb_macros(cls._duckdb_client)
+
+        tables_exist = cls._tables_exist()
+
+        if not tables_exist:
+            if not cls._enable_fetching:
+                raise RuntimeError(
+                    dedent(
+                        """Intermediate Model Test Utils Error:
+                    
+                    - Input test data has not been fetched yet.
+                    
+                    To resolve this error manually set `_enable_fetching = True`
+                    on your test class.
+                    """
+                    )
+                )
+
+            else:
+                log.info("Fetching test data from GCS.")
+                cls._fetch_test_data()
+                log.info("Fetched test data from GCS.")
+        else:
+            log.info(f"Using local test data from: {db_path}")
+
+        log.info("Executing model...")
+        model = REGISTERED_INTERMEDIATE_MODELS["daily_address_summary"]
```

Comment:
should the model to be tested as a variable?

Reply:
yeah, this helps make sure that the model function is correctly registered with that name.  When running the model transformation in production the command line invocation uses that same string. For example:
```
$ uv run python opdata chains intermediate_models op daily_address_summary @20241001:+1 --read_from gcs --write-to gcs
```